File: benchmarks/tsb/bench_dropna.ts
/**
 * Benchmark: dropna on Series and DataFrame (axis=0 and axis=1)
 */
import { Series, DataFrame, dropna, dropnaDataFrame } from "../../src/index.js";

const ROWS = 100_000;
const WARMUP = 3;
const ITERATIONS = 10;

// ~10% missing values
const seriesData = Float64Array.from({ length: ROWS }, (_, i) =>
  i % 10 === 0 ? NaN : i * 1.1,
);
const s = new Series(seriesData);

const colA = Float64Array.from({ length: ROWS }, (_, i) => (i % 7 === 0 ? NaN : i * 0.5));
const colB = Float64Array.from({ length: ROWS }, (_, i) => (i % 11 === 0 ? NaN : i * 1.5));
const colC = Float64Array.from({ length: ROWS }, (_, i) => (i % 13 === 0 ? NaN : i * 2.0));
const df = DataFrame.fromColumns({ a: colA, b: colB, c: colC });

for (let i = 0; i < WARMUP; i++) {
  dropna(s);
  dropnaDataFrame(df, { how: "any" });
  dropnaDataFrame(df, { how: "all" });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  dropna(s);
  dropnaDataFrame(df, { how: "any" });
  dropnaDataFrame(df, { how: "all" });
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "dropna",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_dropna.py
"""
Benchmark: dropna on Series and DataFrame (axis=0, how=any, how=all)
"""
import json
import time
import numpy as np
import pandas as pd

ROWS = 100_000
WARMUP = 3
ITERATIONS = 10

rng = np.random.default_rng(42)
series_data = rng.standard_normal(ROWS)
series_data[::10] = np.nan
s = pd.Series(series_data)

col_a = rng.standard_normal(ROWS)
col_b = rng.standard_normal(ROWS)
col_c = rng.standard_normal(ROWS)
col_a[::7] = np.nan
col_b[::11] = np.nan
col_c[::13] = np.nan
df = pd.DataFrame({"a": col_a, "b": col_b, "c": col_c})

for _ in range(WARMUP):
    s.dropna()
    df.dropna(how="any")
    df.dropna(how="all")

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.dropna()
    df.dropna(how="any")
    df.dropna(how="all")
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "dropna",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
